enhancements/hypershift/konnectivity-socks5-proxy.md

```diff
@@ -0,0 +1,124 @@
+status: implementable
+see-also:
+  - "/enhancements/update/ibm-public-cloud-support.md"
+replaces:
+  -
+superseded-by:
+  -
+---
+
+# Konnectivity Socks5 Proxy
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide [HyperShift](https://github.com/openshift/hypershift/) control plane services the ability to establish [GRPC](https://grpc.io/) connections with servers running on guest clusters.
+
+## Motivation
+
+The HyperShift solution deploys OpenShift clusters by hosting their control plane as pods in a central Management Cluster. This type of setup allows the hosting of many cluster control planes on a single Kubernetes cluster. Each Control Plane has a related Guest Cluster where user workloads run.
+
+The [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) is a member of the control plane and runs within the Management Cluster. One requirement for OLM is to support Custom CatalogSources defined by users on the Guest Cluster.
+To meet this requirement, OLM needs to have the ability to establish GRPC connections with CatalogSource services running on the Guest Cluster.
+
+As of today, OLM is unable to connect to CatalogSources defined on the Guest Cluster because:
+- The Services defined on the guest clusters cannot be resolved by the DNS Service on the Management Cluster.
+- OLM cannot connect to CatalogSources running on the Guest Cluster.
+
+For OLM to support custom Catalog Sources it must be able to establish connections with CatalogSources running on the Guest Cluster.
+
+### Goals
+
+The primary goal of this enhancement is to propose a solution that allows services running on the control plane to:
+- Resolve domain names using the DNS service on the Guest Cluster
+- Tunnel requests to services running on the Guest Cluster
+
+### Non-Goals
+
+- N/A
+
+## Proposal
+
+This enhancement proposes the introduction of a [Socks5 proxy](https://en.wikipedia.org/wiki/SOCKS) that forwards requests from the control plane to the guest cluster.
+
+The Control Plane already includes a [Konnectivity Server](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/) which can be used to tunnel requests to IP addresses running on the Guest Cluster. Rather than requiring each component to support establishing a connection with the Konnectivity Server, this enhancement proposes the introduction of a Socks5 proxy that:
+- Runs as a sidecar on containers
+- Establishes a secure connection with the Konnectivity Server
+- Resolves Domain Names using the DNS Service on Guest Clusters
```

Comment:
Hmm, this reminds me that service name and IPs are all in the service resources in the kube apiserver. A possibility would be not to mess with DNS at all and simply resolve names by querying kube apiserver services.

Reply:
We still need DNS or not? The OLM tries to talk to a kube service which is DNS name pointing to a virtual IP realized by kube-proxy. Hypothetically we could also run kube-proxy alongside the other controlplane so the traffic needs one hop less inside the cluster.

Reply:
What I meant is that typically you'll get a url like: `[service-name].[namespace].svc` 
Instead of doing a DNS lookup, you could simply look up the Service resource with name `[service-name]` under namespace `[namespace]` and get its IP address.

Reply:
true, that is basically what kube-dns does and then provide it through DNS :) I agree we could tell individual components to implement such a lookup, but it seems simpler to re-use the existing implementation in kube-dns

Reply:
That's reasonable - we could check if the domain name is known and if not attempt to resolve it by inspecting the service on the k8s cluster. Thoughts?

Reply:
For this one case, I think doing the service lookup will be a lot simpler than the dns route. Simply because you have to either run  an additional core dns outside of the guest cluster, or go through konnectivity to do lookup inside the guest cluster.